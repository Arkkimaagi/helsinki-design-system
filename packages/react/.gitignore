*.css.d.ts
storybook-static