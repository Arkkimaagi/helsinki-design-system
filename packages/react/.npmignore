.*
**/tsconfig.json
**/rollup.config.js
**/jest.config.js
scripts
src
storybook-static
