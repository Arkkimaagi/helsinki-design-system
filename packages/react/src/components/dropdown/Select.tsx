import React, { useRef } from 'react';
import { useSelect, useMultipleSelection } from 'downshift';
import isEqual from 'lodash.isequal';
import uniqueId from 'lodash.uniqueid';

import 'hds-core';

import styles from './Select.module.scss';
import { FieldLabel } from '../../internal/field-label/FieldLabel';
import classNames from '../../utils/classNames';
import { IconAngleDown, IconCheck, IconCrossCircle } from '../../icons';
import { Tag } from '../tag';

export type SelectProps<OptionType> = {
  /**
   * Additional class names to apply to the select
   */
  className?: string;
  /**
   * If `true`, the dropdown will be disabled
   */
  disabled?: boolean;
  /**
   * A helper text that will be shown below the dropdown
   */
  helper?: React.ReactNode;
  /**
   * Used to generate the first part of the id on the elements.
   */
  id?: string;
  /**
   * If `true`, the input and `helper` will be displayed in an invalid state
   */
  invalid?: boolean;
  /**
   * A function used to detect whether an option is disabled ([example](/?path=/story/components-dropdown--disabled-options))
   */
  isOptionDisabled?: (option: OptionType, index: number) => boolean;
  /**
   * The label for the dropdown
   */
  label: React.ReactNode;
  /**
   * Callback fired when the state is changed
   */
  onChange?: <T = unknown>(selectedItem: T) => void;
  /**
   * Sets the data item field that represents the item label.
   * E.g. an `optionLabelField` value of `'foo'` and a data item `{ foo: 'Label', bar: 'value' }`, would display `Label` in the menu for that specific item
   */
  optionLabelField?: string;
  /**
   * Array of options that should be shown in the menu
   */
  options: OptionType[];
  /**
   * Short hint displayed in the dropdown before the user enters a value
   */
  placeholder?: string;
  /**
   * Label for selected items that is only visible to screen readers. Can be used to to give screen reader users additional information about the selected item.
   */
  selectedItemSrLabel?: string;
  /**
   * Override or extend the root styles applied to the component
   */
  style?: React.CSSProperties;
  /**
   * Sets the number of options that are visible in the menu before it becomes scrollable
   */
  visibleOptions?: number;
} & MultiselectProps<OptionType>;

type MultiselectProps<OptionType> =
  | {
      multiselect?: false;
      value?: OptionType;
      clearButtonAriaLabel?: string;
      selectedItemRemoveButtonAriaLabel?: string;
    }
  | {
      /**
       * Enables selecting multiple values if `true`.
       */
      multiselect?: boolean;
      /**
       * The selected value(s)
       */
      value?: OptionType[];
      /**
       * The aria-label for the clear button
       */
      clearButtonAriaLabel: string;
      /**
       * The aria-label for the selected item remove button
       */
      selectedItemRemoveButtonAriaLabel: string;
    };

/**
 * Helper that checks if an item is in the selected options
 * @param selectedOptions Currently selected options
 * @param item            Item we want to check
 */
function getIsInSelectedOptions<T>(selectedOptions: T[], item: T): boolean {
  return selectedOptions.some((selectedOption: T) => isEqual(selectedOption, item));
}

export const Select = <OptionType,>({
  className,
  clearButtonAriaLabel,
  disabled = false,
  helper,
  id = uniqueId('hds-select-'),
  invalid = false,
  isOptionDisabled,
  label,
  multiselect,
  onChange = () => null,
  // test using options without a label key and without optionLabelField defined
  optionLabelField = 'label',
  options = [],
  placeholder,
  selectedItemRemoveButtonAriaLabel,
  selectedItemSrLabel,
  style,
  value,
  visibleOptions = 5,
}: SelectProps<OptionType>) => {
  const selectedItemsContainerRef = useRef(null);

  // init multi-select
  const {
    activeIndex,
    addSelectedItem,
    getDropdownProps,
    getSelectedItemProps,
    removeSelectedItem,
    reset,
    selectedItems,
    setSelectedItems,
  } = useMultipleSelection<OptionType>({
    // todo: create a prop for setting the removal message
    getA11yRemovalMessage({ itemToString, removedSelectedItem }) {
      console.log(
        `getA11yRemovalMessage message: "${itemToString(removedSelectedItem[optionLabelField])} has been removed"`,
      );
      return `${itemToString(removedSelectedItem[optionLabelField])} has been removed`;
    },
    // sets focus on the first selected item when the select is initialized
    defaultActiveIndex: 0,
    initialActiveIndex: 0,
    // todo: remove. just for testing
    initialSelectedItems: [options[0], options[1]],
    onSelectedItemsChange({ selectedItems: _selectedItems }) {
      return multiselect && onChange(_selectedItems);
    },
    ...(multiselect && value !== undefined && { selectedItems: (value as OptionType[]) ?? [] }),
  });

  // init select
  const {
    getItemProps,
    getLabelProps,
    getMenuProps,
    getToggleButtonProps,
    highlightedIndex,
    isOpen,
    selectedItem,
    selectItem,
  } = useSelect<OptionType>({
    // todo: remove. for testing only.
    // isOpen: true,
    // todo: create a prop for setting the selection message and "selections cleared" message
    // todo: how can this be done for multiselect?
    getA11ySelectionMessage({ selectedItem: _selectedItem }) {
      if (!multiselect && _selectedItem) {
        const message = `${_selectedItem?.[optionLabelField]} has been selected`;
        console.log(`getA11ySelectionMessage message: "${message}"`);
        return message;
      }
      return '';
    },
    id,
    items: options,
    itemToString: (item): string => (item ? item[optionLabelField] ?? '' : ''),
    onSelectedItemChange: ({ selectedItem: _selectedItem }) => !multiselect && onChange(_selectedItem),
    // a defined value indicates that the dropdown should be controlled
    // don't set selectedItem if it's not, so that downshift can handle the state
    ...(!multiselect && value !== undefined && { selectedItem: value as OptionType }),
    stateReducer(state, { type, changes }) {
      switch (type) {
        case useSelect.stateChangeTypes.MenuKeyDownEnter:
        case useSelect.stateChangeTypes.MenuKeyDownSpaceButton:
        case useSelect.stateChangeTypes.ItemClick:
          // prevent the menu from being closed when the user selects an item
          if (multiselect) {
            return {
              ...changes,
              isOpen: state.isOpen,
              highlightedIndex: state.highlightedIndex,
            };
          }
          return changes;
        default:
          return changes;
      }
    },
    onStateChange({ type, selectedItem: _selectedItem }) {
      switch (type) {
        case useSelect.stateChangeTypes.MenuKeyDownEnter:
        case useSelect.stateChangeTypes.MenuKeyDownSpaceButton:
        case useSelect.stateChangeTypes.ItemClick:
        case useSelect.stateChangeTypes.MenuBlur:
          if (multiselect && _selectedItem) {
            getIsInSelectedOptions(selectedItems, _selectedItem)
              ? setSelectedItems(selectedItems.filter((item) => !isEqual(item, _selectedItem)))
              : addSelectedItem(_selectedItem);
            selectItem(null);
          }
          break;
        default:
          break;
      }
    },
  });

  if (!multiselect) {
    // we call the getDropdownProps getter function when multiselect isn't enabled
    // in order to suppress the "You forgot to call the ..." error message thrown by downshift.
    // we only need to apply the getter props to the toggle button when multiselect is enabled.
    getDropdownProps({}, { suppressRefError: true });
  }

  // returns the toggle button label based on the dropdown mode
  const getButtonLabel = (): React.ReactNode => {
    let buttonLabel = selectedItem?.[optionLabelField] || placeholder;
    if (multiselect) buttonLabel = selectedItems.length > 0 ? null : placeholder;
    return buttonLabel;
  };
  // show placeholder if no value is selected
  const showPlaceholder = (multiselect && selectedItems.length === 0) || (!multiselect && !selectedItem);

  return (
    <div
      className={classNames(
        styles.root,
        invalid && styles.invalid,
        disabled && styles.disabled,
        isOpen && styles.open,
        multiselect && styles.multiselect,
        className,
      )}
      style={style}
    >
      {/* LABEL */}
      {label && <FieldLabel label={label} {...getLabelProps()} />}
      <div className={styles.wrapper}>
        {multiselect && selectedItems.length > 0 && (
          <>
            {/* SELECTED ITEMS */}
            <div ref={selectedItemsContainerRef} className={styles.selectedItems}>
              {selectedItems.map((_selectedItem, index) => {
                const selectedItemLabel = _selectedItem[optionLabelField];
                const tagId = uniqueId('hds-tag-');

                return (
                  <Tag
                    key={selectedItemLabel}
                    className={styles.tag}
                    id={tagId}
                    label={selectedItemLabel}
                    labelProps={{ 'aria-labelledby': `${id}-label ${tagId}-label` }}
                    deleteButtonAriaLabel={selectedItemRemoveButtonAriaLabel}
                    onDelete={(e) => {
                      e.stopPropagation();
                      removeSelectedItem(_selectedItem);
                    }}
                    srOnlyLabel={selectedItemSrLabel}
                    {...getSelectedItemProps({ selectedItem: _selectedItem, index })}
                  />
                );
              })}
            </div>
            {/* CLEAR BUTTON */}
            <button
              type="button"
              className={styles.clearButton}
              onClick={() => reset()}
              aria-label={clearButtonAriaLabel}
              onFocus={() => {
                // manually set the tabindex of the first selected item to "0"
                // when the clear button is focused and Downshift's activeIndex is -1,
                // otherwise shift+tabbing back to the selected items won't work as they have a tabindex value of -1.
                if (activeIndex === -1) {
                  // eslint-disable-next-line no-unused-expressions
                  selectedItemsContainerRef?.current.childNodes[0].setAttribute('tabindex', '0');
                }
              }}
            >
              <IconCrossCircle />
            </button>
          </>
        )}
        {/* TOGGLE BUTTON */}
        <button
          type="button"
          {...getToggleButtonProps({
            'aria-owns': getMenuProps().id,
            // prepend helper text id to the id's return by the downshift getter function,
            // so that the helper text will be read to screen reader users before the other labels.
            // todo: only add helper id if a helper text is defined
            'aria-labelledby': `${id}-helper ${getToggleButtonProps()['aria-labelledby']}`,
            // add downshift dropdown props when multiselect is enabled
            ...(multiselect && { ...getDropdownProps({ preventKeyAction: isOpen }) }),
            disabled,
            className: classNames(styles.button, showPlaceholder && styles.placeholder),
          })}
        >
          {getButtonLabel()}
          <IconAngleDown className={styles.angleIcon} />
        </button>
        {/* INVALID ICON */}
        {/* {invalid && <IconAlertCircle className={styles.invalidIcon} />} */}
        <ul
          {...getMenuProps({
            ...(multiselect && { 'aria-multiselectable': true }),
            className: classNames(styles.menu, options.length > visibleOptions && styles.overflow),
            style: { maxHeight: `calc(var(--menu-item-height) * ${visibleOptions})` },
          })}
        >
          {isOpen &&
            options.map((item, index) => {
              const optionLabel = item[optionLabelField];
              const selected = multiselect ? getIsInSelectedOptions(selectedItems, item) : isEqual(selectedItem, item);
              // todo: add aria-disabled to disabled menu items
              const optionDisabled = typeof isOptionDisabled === 'function' ? isOptionDisabled(item, index) : false;

              return (
                <li
                  key={optionLabel}
                  {...getItemProps({
                    item,
                    index,
                    disabled: optionDisabled,
                    className: classNames(
                      styles.menuItem,
                      highlightedIndex === index && styles.highlighted,
                      selected && styles.selected,
                      optionDisabled && styles.disabled,
                      // todo: use root multiselect class
                      multiselect && styles.multiselect,
                    ),
                  })}
                  {...{ 'aria-selected': selected }}
                >
                  {multiselect ? (
                    <>
                      <IconCheck className={styles.checkbox} aria-hidden />
                      {optionLabel}
                    </>
                  ) : (
                    <>
                      {optionLabel}
                      {selected && <IconCheck className={styles.selectedIcon} />}
                    </>
                  )}
                </li>
              );
            })}
        </ul>
      </div>
      {helper && (
        <div id={`${id}-helper`} className={styles.helperText} aria-hidden="true">
          {helper}
        </div>
      )}
    </div>
  );
};
