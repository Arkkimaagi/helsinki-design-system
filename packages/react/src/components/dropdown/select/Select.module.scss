// todo: test theming. selected icon, checkbox etc.
@import "../dropdown";

.root {
  @extend %dropdownRoot;
}

/**
 * WRAPPER
 */

.wrapper {
  @extend %dropdownWrapper;
}

/**
 * BUTTON
 */

// TODO: USE BUTTON RESET PLACEHOLDER CLASS WHEN FOOTER IS MERGED
.button {
  @extend %dropdownToggleButton;
}
//.button {
//  background: none;
//  border: none;
//  color: inherit;
//  cursor: pointer;
//  font: inherit;
//  padding: 0;
//
//  align-items: center;
//  color: var(--dropdown-color-default);
//  display: flex;
//  font-size: var(--fontsize-body-l);
//  outline: none;
//  padding: calc(var(--spacing-s) - var(--border-width) * 2) var(--spacing-s);
//  width: 100%;
//
//  &.placeholder {
//    color: var(--placeholder-color);
//  }
//}

/**
 * ICON
 */

.icon {
  @extend %dropdownIcon;
}

/**
 * ANGLE ICON
 */

.angleIcon {
  @extend %dropdownAngleIcon;
}

/**
 * MENU
 */

.menu {
  @extend %dropdownMenu;
}

/**
 * MENU ITEM
 */

.menuItem {
  @extend %dropdownMenuItem;
}

/**
 * HELPER TEXT
 */

.helperText {
  @extend %dropdownHelperText;
}

/**
 * DISABLED
 */

.disabled {
  @extend %dropdownDisabled;
}

/**
 * INVALID
 */

.invalid {
  @extend %dropdownInvalid;
}

/**
 * MULTISELECT
 */

.multiselect {
  @extend %dropdownMultiselect;
}
