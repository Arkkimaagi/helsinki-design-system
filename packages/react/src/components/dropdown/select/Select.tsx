import React, { FocusEvent, useCallback, useRef, useState } from 'react';
import {
  useSelect,
  useMultipleSelection,
  A11yRemovalMessage,
  A11yStatusMessageOptions,
  UseMultipleSelectionStateChange,
  UseMultipleSelectionStateChangeOptions,
  UseMultipleSelectionState,
} from 'downshift';
import isEqual from 'lodash.isequal';
import uniqueId from 'lodash.uniqueid';
import { useVirtual } from 'react-virtual';

import 'hds-core';

import styles from './Select.module.scss';
import { FieldLabel } from '../../../internal/field-label/FieldLabel';
import classNames from '../../../utils/classNames';
import { IconAlertCircle, IconAngleDown } from '../../../icons';
import { SelectedItems } from '../../../internal/selectedItems/SelectedItems';
import {
  DROPDOWN_MENU_ITEM_HEIGHT,
  getIsElementBlurred,
  getIsElementFocused,
  getIsInSelectedOptions,
} from '../dropdownUtils';
import { DropdownMenu } from '../dropdownMenu/DropdownMenu';

export type SelectProps<OptionType> = {
  /**
   * When set to `true`, allows moving from the first item to the last item with Arrow Up, and vice versa using Arrow Down.
   */
  circularNavigation?: boolean;
  /**
   * Additional class names to apply to the select
   */
  className?: string;
  /**
   * Flag for whether the clear selections button should be displayed
   */
  clearable?: boolean;
  /**
   * If `true`, the dropdown will be disabled
   */
  disabled?: boolean;
  /**
   * Function used to generate an ARIA a11y message when the status changes. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useSelect#geta11ystatusmessage) for more information.
   */
  getA11yStatusMessage?: (options: A11yStatusMessageOptions<OptionType>) => string;
  /**
   * A helper text that will be shown below the dropdown
   */
  helper?: React.ReactNode;
  /**
   * An error text that will be shown below the dropdown when `invalid` is true
   */
  error?: React.ReactNode;
  /**
   * Used to generate the first part of the id on the elements
   */
  id?: string;
  /**
   * If `true`, the input and `helper` will be displayed in an invalid state
   */
  invalid?: boolean;
  /**
   * A function used to detect whether an option is disabled
   */
  isOptionDisabled?: (option: OptionType, index: number) => boolean;
  /**
   * The label for the dropdown
   */
  label: React.ReactNode;
  /**
   * Callback function fired when the component is blurred
   */
  onBlur?: () => void;
  /**
   * Callback function fired when the state is changed
   */
  onChange?: (selected: OptionType | OptionType[]) => void;
  /**
   * Callback function fired when the component is focused
   */
  onFocus?: () => void;
  /**
   * Sets the data item field that represents the item label
   * E.g. an `optionLabelField` value of `'foo'` and a data item `{ foo: 'Label', bar: 'value' }`, would display `Label` in the menu for that specific item
   */
  optionLabelField?: string;
  /**
   * Array of options that should be shown in the menu
   */
  options: OptionType[];
  /**
   * Short hint displayed in the dropdown before the user enters a value
   */
  placeholder?: string;
  /**
   * If `true`, marks the dropdown as required
   */
  required?: boolean;
  /**
   * Label for selected items that is only visible to screen readers. Can be used to to give screen reader users additional information about the selected item
   */
  selectedItemSrLabel?: string;
  /**
   * Override or extend the root styles applied to the component
   */
  style?: React.CSSProperties;
  /**
   * todo
   */
  virtualized?: boolean;
  /**
   * Sets the number of options that are visible in the menu before it becomes scrollable
   */
  visibleOptions?: number;
} & MultiselectProps<OptionType>;

type MultiselectProps<OptionType> =
  | {
      multiselect?: false;
      clearButtonAriaLabel?: string;
      defaultValue?: OptionType;
      getA11yRemovalMessage?: undefined;
      getA11ySelectionMessage?: (options: A11yStatusMessageOptions<OptionType>) => string;
      icon?: React.ReactNode;
      selectedItemRemoveButtonAriaLabel?: string;
      value?: OptionType;
    }
  | {
      /**
       * Enables selecting multiple values if `true`.
       */
      multiselect?: boolean;
      /**
       * The aria-label for the clear button
       */
      clearButtonAriaLabel: string;
      /**
       * Value(s) that should be selected when the dropdown is initialized
       */
      defaultValue?: OptionType[];
      /**
       * Function used to generate an ARIA a11y message when an item is removed. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useMultipleSelection#geta11yremovalmessage) for more information.
       */
      getA11yRemovalMessage?: (options: A11yRemovalMessage<OptionType>) => string;
      /**
       * Function used to generate an ARIA a11y message when an item is selected. See [here](https://github.com/downshift-js/downshift/tree/master/src/hooks/useSelect#geta11yselectionmessage) for more information.
       */
      getA11ySelectionMessage?: undefined;
      /**
       * Icon to be shown in the dropdown
       */
      icon?: undefined;
      /**
       * The aria-label for the selected item remove button
       */
      selectedItemRemoveButtonAriaLabel: string;
      /**
       * The selected value(s)
       */
      value?: OptionType[];
    };

/**
 * Multi-select state change handler
 * @param type
 * @param activeIndex
 * @param currentActiveIndex
 * @param selectedItemsContainerEl
 */
export function onMultiSelectStateChange<T>(
  { type, activeIndex }: UseMultipleSelectionStateChange<T>,
  currentActiveIndex,
  selectedItemsContainerEl,
) {
  let activeNode;
  const { FunctionRemoveSelectedItem, SelectedItemKeyDownBackspace } = useMultipleSelection.stateChangeTypes;

  if (type === FunctionRemoveSelectedItem || type === SelectedItemKeyDownBackspace) {
    activeNode = selectedItemsContainerEl?.childNodes[currentActiveIndex] as HTMLDivElement;
    if (!activeIndex && activeNode) activeNode.focus();
  }
}

/**
 * Multi-select reducer function
 * @param state
 * @param type
 * @param changes
 * @param controlled
 */
export function multiSelectReducer<T>(
  state: UseMultipleSelectionState<T>,
  { type, changes }: UseMultipleSelectionStateChangeOptions<T>,
  controlled: boolean,
) {
  const { FunctionRemoveSelectedItem, SelectedItemKeyDownBackspace } = useMultipleSelection.stateChangeTypes;

  if (type === FunctionRemoveSelectedItem || type === SelectedItemKeyDownBackspace) {
    // get the index of the deleted item
    const removedItemIndex = state.selectedItems.findIndex((item) => !changes.selectedItems.includes(item));
    // activeIndex updates at a different time depending on whether the
    // component is controlled. For controlled components, when 'onStateChange'
    // is called, the removed item still has a SelectedItem representing
    // it in the DOM. This means that the succeeding SelectedItem is at
    // index n + 1 instead of n
    const adjustedIndex = controlled ? removedItemIndex + 1 : removedItemIndex;
    // whether the removed item was last in selectedItems
    const lastItemRemoved = removedItemIndex === changes.selectedItems.length;

    return {
      ...changes,
      // set the new last item as active if the removed item was last,
      // otherwise set the item succeeding the removed one active
      activeIndex: lastItemRemoved ? removedItemIndex - 1 : adjustedIndex,
    };
  }

  return changes;
}

export const Select = <OptionType,>({
  circularNavigation = false,
  className,
  clearable = true,
  clearButtonAriaLabel,
  defaultValue,
  disabled = false,
  error,
  getA11yRemovalMessage = () => '',
  getA11ySelectionMessage = () => '',
  getA11yStatusMessage = () => '',
  helper,
  icon,
  id = uniqueId('hds-select-'),
  invalid = false,
  isOptionDisabled,
  label,
  multiselect = false,
  onBlur = () => null,
  onChange = () => null,
  onFocus = () => null,
  optionLabelField = 'label',
  options = [],
  placeholder,
  required,
  selectedItemRemoveButtonAriaLabel,
  selectedItemSrLabel,
  style,
  value,
  virtualized = false,
  visibleOptions = 5,
}: SelectProps<OptionType>) => {
  // flag for whether the component is controlled
  const controlled = multiselect && value !== undefined;
  // selected items container ref
  const selectedItemsContainerRef = useRef<HTMLDivElement>(null);
  // menu ref
  const menuRef = React.useRef<HTMLUListElement>();
  // whether active focus is within the dropdown
  const [hasFocus, setFocus] = useState(false);
  // virtualize menu items to increase performance
  const virtualizer = useVirtual<HTMLUListElement>({
    size: options.length,
    parentRef: menuRef,
    estimateSize: useCallback(() => DROPDOWN_MENU_ITEM_HEIGHT, []),
    overscan: visibleOptions,
  });

  // init multi-select
  const {
    activeIndex,
    addSelectedItem,
    getDropdownProps,
    getSelectedItemProps,
    removeSelectedItem,
    reset,
    selectedItems,
    setActiveIndex,
    setSelectedItems,
  } = useMultipleSelection<OptionType>({
    // sets focus on the first selected item when the dropdown is initialized
    defaultActiveIndex: 0,
    initialActiveIndex: 0,
    // set the default value(s) when the dropdown is initialized
    initialSelectedItems: (defaultValue as OptionType[]) ?? [],
    // set the selected items when the dropdown is controlled
    ...(controlled && { selectedItems: (value as OptionType[]) ?? [] }),
    getA11yRemovalMessage,
    onSelectedItemsChange: ({ selectedItems: _selectedItems }) => multiselect && onChange(_selectedItems),
    onStateChange: (changes) =>
      onMultiSelectStateChange<OptionType>(changes, activeIndex, selectedItemsContainerRef.current),
    stateReducer: (state, actionAndChanges) => multiSelectReducer<OptionType>(state, actionAndChanges, controlled),
  });

  // init select
  const {
    getItemProps,
    getLabelProps,
    getMenuProps,
    getToggleButtonProps,
    highlightedIndex,
    isOpen,
    selectedItem,
    selectItem,
  } = useSelect<OptionType>({
    circularNavigation,
    id,
    items: options,
    // set the default value when the dropdown is initialized
    initialSelectedItem: (defaultValue as OptionType) ?? null,
    // a defined value indicates that the dropdown should be controlled
    // don't set selectedItem if it's not, so that downshift can handle the state
    ...(!multiselect && value !== undefined && { selectedItem: value as OptionType }),
    getA11ySelectionMessage,
    getA11yStatusMessage,
    itemToString: (item): string => (item ? item[optionLabelField] ?? '' : ''),
    onSelectedItemChange: ({ selectedItem: _selectedItem }) => !multiselect && onChange(_selectedItem),
    onStateChange({ type, selectedItem: _selectedItem }) {
      const { ItemClick, MenuBlur, MenuKeyDownEnter, MenuKeyDownSpaceButton } = useSelect.stateChangeTypes;

      if (
        (type === ItemClick || type === MenuBlur || type === MenuKeyDownEnter || type === MenuKeyDownSpaceButton) &&
        multiselect &&
        _selectedItem
      ) {
        getIsInSelectedOptions(selectedItems, _selectedItem)
          ? setSelectedItems(selectedItems.filter((item) => !isEqual(item, _selectedItem)))
          : addSelectedItem(_selectedItem);
        selectItem(null);
      }
    },
    stateReducer(state, { type, changes }) {
      const { ItemClick, MenuKeyDownSpaceButton } = useSelect.stateChangeTypes;

      // prevent the menu from being closed when the user selects an item by clicking or pressing space
      if ((type === ItemClick || type === MenuKeyDownSpaceButton) && multiselect) {
        return {
          ...changes,
          isOpen: state.isOpen,
          highlightedIndex: state.highlightedIndex,
        };
      }

      return changes;
    },
  });

  const handleWrapperFocus = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementFocused(e)) {
      setFocus(true);
      onFocus();
    }
  };

  const handleWrapperBlur = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementBlurred(e)) {
      setFocus(false);
      onBlur();
    }
  };

  if (!multiselect) {
    // we call the getDropdownProps getter function when multiselect isn't enabled
    // in order to suppress the "You forgot to call the ..." error message thrown by downshift.
    // we only need to apply the getter props to the toggle button when multiselect is enabled.
    getDropdownProps({}, { suppressRefError: true });
  }

  // returns the toggle button label based on the dropdown mode
  const getButtonLabel = (): React.ReactNode => {
    let buttonLabel = selectedItem?.[optionLabelField] || placeholder;
    if (multiselect) buttonLabel = selectedItems.length > 0 ? null : placeholder;
    return buttonLabel;
  };

  // screen readers should read the labels in the following order:
  // field label > helper text > error text > toggle button label
  // helper and error texts should only be read if they have been defined
  // prettier-ignore
  const buttonAriaLabel =
    `${getLabelProps().id}${error ? ` ${id}-error` : ''}${helper ? ` ${id}-helper` : ''} ${getToggleButtonProps().id}`;

  // show placeholder if no value is selected
  const showPlaceholder = (multiselect && selectedItems.length === 0) || (!multiselect && !selectedItem);

  return (
    <div
      className={classNames(
        styles.root,
        invalid && styles.invalid,
        disabled && styles.disabled,
        isOpen && styles.open,
        multiselect && styles.multiselect,
        className,
      )}
      style={style}
    >
      {/* LABEL */}
      {label && <FieldLabel label={label} required={required} {...getLabelProps()} />}
      <div className={styles.wrapper} onFocus={handleWrapperFocus} onBlur={handleWrapperBlur}>
        {/* SELECTED ITEMS */}
        {multiselect && selectedItems.length > 0 && (
          <SelectedItems
            activeIndex={activeIndex}
            clearable={clearable}
            clearButtonAriaLabel={clearButtonAriaLabel}
            dropdownId={id}
            getSelectedItemProps={getSelectedItemProps}
            hideItems={!hasFocus}
            onClear={() => reset()}
            onRemove={removeSelectedItem}
            optionLabelField={optionLabelField}
            removeButtonAriaLabel={selectedItemRemoveButtonAriaLabel}
            selectedItems={selectedItems}
            selectedItemSrLabel={selectedItemSrLabel}
            selectedItemsContainerRef={selectedItemsContainerRef}
            setActiveIndex={setActiveIndex}
          />
        )}
        {/* TOGGLE BUTTON */}
        <button
          type="button"
          {...getToggleButtonProps({
            'aria-owns': getMenuProps().id,
            'aria-labelledby': buttonAriaLabel,
            // add downshift dropdown props when multiselect is enabled
            ...(multiselect && { ...getDropdownProps({ preventKeyAction: isOpen }) }),
            ...(invalid && { 'aria-invalid': true }),
            disabled,
            className: classNames(styles.button, showPlaceholder && styles.placeholder),
          })}
        >
          {/* icons are only supported by single selects */}
          {icon && !multiselect && (
            <span className={styles.icon} aria-hidden>
              {icon}
            </span>
          )}
          {getButtonLabel()}
          <IconAngleDown className={styles.angleIcon} />
        </button>
        {/* MENU */}
        <DropdownMenu
          getItemProps={(optionDisabled, index, item, selected, virtualRow) =>
            getItemProps({
              item,
              index,
              disabled: optionDisabled,
              className: classNames(
                styles.menuItem,
                highlightedIndex === index && styles.highlighted,
                selected && styles.selected,
                optionDisabled && styles.disabled,
                virtualized && styles.virtualized,
              ),
              // apply styles for virtualization to menu items
              ...(virtualRow && {
                style: {
                  transform: `translateY(${virtualRow.start}px`,
                },
                ref: virtualRow.measureRef,
              }),
            })
          }
          isOptionDisabled={isOptionDisabled}
          menuProps={getMenuProps({
            ...(multiselect && { 'aria-multiselectable': true }),
            ...(required && { 'aria-required': true }),
            style: { maxHeight: `calc(var(--menu-item-height) * ${visibleOptions})` },
            ref: menuRef,
          })}
          menuStyles={styles}
          multiselect={multiselect}
          open={isOpen}
          optionLabelField={optionLabelField}
          options={options}
          selectedItem={selectedItem}
          selectedItems={selectedItems}
          virtualizer={virtualized && virtualizer}
          visibleOptions={visibleOptions}
        />
      </div>
      {/* INVALID TEXT */}
      {invalid && error && (
        <div id={`${id}-error`} className={styles.errorText} aria-hidden>
          <IconAlertCircle className={styles.invalidIcon} />
          {error}
        </div>
      )}
      {/* HELPER TEXT */}
      {helper && (
        <div id={`${id}-helper`} className={styles.helperText} aria-hidden>
          {helper}
        </div>
      )}
    </div>
  );
};
