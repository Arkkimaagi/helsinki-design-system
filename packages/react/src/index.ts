// import core base styles
// import 'hds-core';

export * from './icons';
export * from './components';
