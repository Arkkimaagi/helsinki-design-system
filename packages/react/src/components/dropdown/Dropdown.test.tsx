import React from 'react';
import { render } from '@testing-library/react';

// import { Dropdown } from './Dropdown';
import { Select } from './select/Select';

const options = [{ value: 'foo' }, { value: 'bar' }, { value: 'baz' }];
const dropdownProps = {
  options,
  label: 'Label',
  helper: 'Helper',
};

// describe('<Dropdown /> spec', () => {
//   it('renders the component', () => {
//     const { asFragment } = render(<Dropdown {...dropdownProps} />);
//     expect(asFragment()).toMatchSnapshot();
//   });
// });

describe('<Select /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<Select {...dropdownProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

// todo: write tests
// https://helsinkisolutionoffice.atlassian.net/browse/HDS-224
