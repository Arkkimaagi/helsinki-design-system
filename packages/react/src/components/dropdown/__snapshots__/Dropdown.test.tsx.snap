// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <label
      class="label "
      for="hds-toggle-button"
      id="hds-label"
    >
      Label
    </label>
    <div
      class="wrapper"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="hds-label hds-toggle-button"
        class="buttonDropdown placeholder"
        id="hds-toggle-button"
        type="button"
      >
        <span
          class="buttonLabel"
        />
        <svg
          class="icon s angleIcon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
    </div>
    <ul
      aria-labelledby="hds-label"
      class="menu"
      id="hds-menu"
      role="listbox"
      style="max-height: calc(var(--dropdown-height) * 5);"
      tabindex="-1"
    />
    <div
      class="helperText"
    >
      Helper
    </div>
  </div>
</DocumentFragment>
`;
